Code/Framework/AzCore/Tests/Math/Vector4Tests.cpp

```diff
@@ -14,448 +14,483 @@
-            angleTest(angleFunc, Vector4{ 42.0f, 0.0f, 0.0f, 0.0f }, Vector4{ 0.0f, 23.0f, 0.0f, 0.0f }, AZ::Constants::HalfPi);
-            angleTest(angleFunc, Vector4{ 1.0f, 0.0f, 0.0f, 0.0f }, Vector4{ -1.0f, 0.0f, 0.0f, 0.0f }, AZ::Constants::Pi);
-            angleTest(angleFunc, Vector4{ 1.0f, 0.0f, 0.0f, 0.0f }, Vector4{ 1.0f, 1.0f, 0.0f, 0.0f }, AZ::Constants::QuarterPi);
-            angleTest(angleFunc, Vector4{ 1.0f, 0.0f, 0.0f, 0.0f }, Vector4{ 1.0f, 0.0f, 0.0f, 0.0f }, 0.f);
-            angleTest(angleFunc, Vector4{ 1.0f, 1.0f, 0.0f, 0.0f }, Vector4{ -1.0f, -1.0f, 0.0f, 0.0f }, AZ::Constants::Pi);
-        }
-
-        const Vec4CalcFunc angleDegFuncs[2] = { &Vector4::AngleDeg, &Vector4::AngleSafeDeg };
-        for (Vec4CalcFunc angleDegFunc : angleDegFuncs)
-        {
-            angleTest(angleDegFunc, Vector4{ 1.0f, 0.0f, 0.0f, 0.0f }, Vector4{ 0.0f, 1.0f, 0.0f, 0.0f }, 90.f);
-            angleTest(angleDegFunc, Vector4{ 42.0f, 0.0f, 0.0f, 0.0f }, Vector4{ 0.0f, 23.0f, 0.0f, 0.0f }, 90.f);
-            angleTest(angleDegFunc, Vector4{ 1.0f, 0.0f, 0.0f, 0.0f }, Vector4{ -1.0f, 0.0f, 0.0f, 0.0f }, 180.f);
-            angleTest(angleDegFunc, Vector4{ 1.0f, 0.0f, 0.0f, 0.0f }, Vector4{ 1.0f, 1.0f, 0.0f, 0.0f }, 45.f);
-            angleTest(angleDegFunc, Vector4{ 1.0f, 0.0f, 0.0f, 0.0f }, Vector4{ 1.0f, 0.0f, 0.0f, 0.0f }, 0.f);
-            angleTest(angleDegFunc, Vector4{ 1.0f, 1.0f, 0.0f, 0.0f }, Vector4{ -1.0f, -1.0f, 0.0f, 0.0f }, 180.f);
-        }
-
-        const Vec4CalcFunc angleSafeFuncs[2] = { &Vector4::AngleSafe, &Vector4::AngleSafeDeg };
-        for (Vec4CalcFunc angleSafeFunc : angleSafeFuncs)
-        {
-            angleTest(angleSafeFunc, Vector4{ 0.0f, 0.0f, 0.0f, 0.0f }, Vector4{ 0.0f, 1.0f, 0.0f, 0.0f }, 0.f);
-            angleTest(angleSafeFunc, Vector4{ 0.0f, 0.0f, 0.0f, 0.0f }, Vector4{ 0.0f, 0.0f, 0.0f, 0.0f }, 0.f);
-            angleTest(angleSafeFunc, Vector4{ 1.0f, 0.0f, 0.0f, 0.0f }, Vector4{ 0.0f, 0.0f, 0.0f, 0.0f }, 0.f);
-            angleTest(angleSafeFunc, Vector4{ 0.0f, 0.0f, 0.0f, 0.0f }, Vector4{ 0.0f, 323432.0f, 0.0f, 0.0f }, 0.f);
-            angleTest(angleSafeFunc, Vector4{ 323432.0f, 0.0f, 0.0f, 0.0f }, Vector4{ 0.0f, 0.0f, 0.0f, 0.0f }, 0.f);
-        }
-    }
-}
+        EXPECT_THAT(v1, IsClose(AZ::Vector4(6.0f, 9.0f, -1.5f, 7.5f)));
+    }
+
+    struct AngleTestArgs
+    {
+        AZ::Vector4 current;
+        AZ::Vector4 target;
+        float angle;
+    };
+
+    using AngleTestFixture = ::testing::TestWithParam<AngleTestArgs>;
+
+    TEST_P(AngleTestFixture, TestAngle)
+    {
+        auto& param = GetParam();
+        EXPECT_FLOAT_EQ(param.current.Angle(param.target), param.angle);
+    }
+
+    TEST_P(AngleTestFixture, TestAngleSafe)
+    {
+        auto& param = GetParam();
+        EXPECT_FLOAT_EQ(param.current.AngleSafe(param.target), param.angle);
+    }
+
+    INSTANTIATE_TEST_CASE_P(
```

Comment:
The macro metalanguage really obfuscates the testing code. Just out of curiosity, what are the benefits of moving from the lambda to macros?

Reply:
the same arguments are fed into the parameterized test for TestAngle and TestAngleSafe. the parameters defined here are separated into individual test cases, easier to see which input fails the test.